### PR TITLE
Add Change Donations to minter

### PR DESCRIPTION
**Pull request by amilz**
**Description:** 
Adds an `Add Charity` Button to minter so that user  can either add a creator per normal OR add a royalty share for a non-profit. OnClick, a `ChangeDonationField` component will be displayed prompting the user to search for a charity. The user can search for a non-profit. The search will return a list of non-profits matching their query by fetching an array of `ChangeNonprofit` from Change API. 
User can `Select` a non-profit, and the non-profit will be added to the Royalty list based on the `.crypto.solana_address` of the selected nonproft.
Otherwise removing creators or modifying a creator is left the same. 

**Summary of Changes**
- Created  `ChangeDonationField` Component modelled off of `RoyaltiesCreators.tsx`'s handling of `Add Creator`
  - Display search bar that queries Change API through `performSearch`
  - On Select, charity is added to Royalty list through `addDonation` (modelled off of `addCreator`)
- Modified `RoyaltiesCreators.tsx` to accomodate Charities: 
  - Add `Add Charity` button
  - Add `ChangeDonationField` Component 
  - Add `(charity)` display for Charity added to royalty list (`src/modules/bulk-minter/components/RoyaltiesCreators.tsx`)
  - Add `(charity)` display for Charity added to royalty list (`src/modules/bulk-minter/components/RoyaltiesCreators.tsx`)
- Add `ChangeNonprofit` Type (`src/utils/change.ts`)
- Add optional `Charity Props` to `Creator` Type (`src/modules/bulk-minter/index.tsx`)
- Add `src/modules/bulk-minter/components/Loading.tsx`, a spinner overlay called when User is searching or the Change API is fetching results. 
- Add a `Debounce` class for handling the search (`src/utils/debounce.ts`)

**Screenshots**
Add Charity Button 
image

Charity Search Bar
image

Initial Search 
image

Found Results
image

Charity Added to Royalty List
image

Already Added Notification (if user searches for same charity that's already been added)
image

### CODE REVIEW

File: src/modules/bulk-minter/components/Loading.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import React, { FunctionComponent } from "react";
+
+interface LoadingProps {
```
Comment (austinmilt) [inline src/modules/bulk-minter/components/Loading.tsx:3]: Just a comment for other reviewers: I think we should save tailwind-ifying everything for a future PR where we overhaul this repo

File: src/modules/bulk-minter/components/RoyaltiesCreators.tsx [hunk 2]
```diff
@@ -227,19 +234,34 @@ const CreatorsRow = ({
...
+            '...' +
+            props.creatorAddress.slice(props.creatorAddress.length - 4)}
       </Paragraph>
-      {!isHolaplex && (
-        <FeatherIcon
-          className="creator-row-icon"
-          icon="copy"
+      {props.charityProps?.isCharity && (
+        <a href={CHANGE_BASE_URL + props.creatorAddress} target="_blank">
+          <FeatherIcon className="creator-row-icon" icon="external-link" />
```
Comment (austinmilt) [inline src/modules/bulk-minter/components/RoyaltiesCreators.tsx:245]:
```suggestion
          <FeatherIcon icon="external-link" />
```

File: src/modules/bulk-minter/components/RoyaltiesCreators.tsx [hunk 3]
```diff
@@ -227,19 +234,34 @@ const CreatorsRow = ({
...
+      {!isHolaplex && !props.charityProps?.isCharity && (
+        <button
           onClick={() => {
-            navigator.clipboard.writeText(creatorAddress);
+            navigator.clipboard.writeText(props.creatorAddress);
             toast('Address copied to clipboard!');
           }}
-        />
+        >
+          <FeatherIcon className="creator-row-icon" icon="copy" />
```
Comment (austinmilt) [inline src/modules/bulk-minter/components/RoyaltiesCreators.tsx:256]:
```suggestion
          <FeatherIcon icon="copy" />
```

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    </div>
+  );
+};
+
+interface Props {
+  showDonationField: boolean;
+  creators: Creator[];
+  updateDonationFieldState: (show: React.SetStateAction<boolean>) => void;
+  updateCreatorState: (value: React.SetStateAction<Creator[]>) => void;
+  updateShowErrorState: (value: React.SetStateAction<boolean>) => void;
```
Comment (austinmilt) [inline src/modules/bulk-minter/components/change/ChangeDonationField.tsx]:
```suggestion
  updateDonationFieldState: (show: boolean) => void;
  updateCreatorState: (value: Creator[]) => void;
  updateShowErrorState: (value: boolean) => void;
```

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        <Row>
+          <Form.Item
+            name="addDonation"
+            style={{ width: '100%', marginBottom: '0px' }}
+            rules={[
+              { required: true, message: 'Please enter a value' },
+              {
+                message: `You can only add ${MAX_CREATOR_LIMIT} creators`,
+                async validator() {
+                  if (props.creators.length === MAX_CREATOR_LIMIT) {
```
Comment (austinmilt) [inline src/modules/bulk-minter/components/change/ChangeDonationField.tsx]: should this be `if (props.creators.length > MAX_CREATOR_LIMIT) {`? If not, would be good to put a comment that says why this is correct

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                >
+                  &nbsp;
+                </Row>
+              )}
+
+              {searchResults.map((result) => {
+                const { solana_address } = result.crypto;
+                const existingCreators = props.creators.map((c) => c.address);
+                const indexOfDuplicate = existingCreators.findIndex((a) => solana_address === a);
+                const enableSelectButton = indexOfDuplicate === -1;
```
Comment (austinmilt) [inline src/modules/bulk-minter/components/change/ChangeDonationField.tsx:259]: Is this actually a duplicate, or just locating a creator that has already been added?

```suggestion
                const creatorAlreadyAdded = existingCreators.findIndex((a) => solana_address === a) !== -1;
                const enableSelectButton = !creatorAlreadyAdded;
```

**Review by austinmilt — APPROVED** [review]
Some minor stuff that may or may not need to be changed. Please check before merging.

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx (hunk 6 above)
Comment (straversi) [reply src/modules/bulk-minter/components/change/ChangeDonationField.tsx:259]: This diff is from a stale commit I believe, can't find that line anymore.

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx (hunk 5 above)
Comment (straversi) [reply src/modules/bulk-minter/components/change/ChangeDonationField.tsx]: I think you're right, this was just copied from an existing check. Updated both the existing one and this new line.

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx (hunk 4 above)
Comment (straversi) [reply src/modules/bulk-minter/components/change/ChangeDonationField.tsx]: ✅

File: src/modules/bulk-minter/components/RoyaltiesCreators.tsx (hunk 3 above)
Comment (straversi) [reply src/modules/bulk-minter/components/RoyaltiesCreators.tsx:256]: These are getting style from `StyledCreatorsRow` on line 66.

File: src/modules/bulk-minter/components/RoyaltiesCreators.tsx (hunk 2 above)
Comment (straversi) [reply src/modules/bulk-minter/components/RoyaltiesCreators.tsx:245]: These are getting style from `StyledCreatorsRow` on line 66.

**Comment by straversi** [comment]
@austinmilt Addressed your latest round of comments!

File: src/modules/bulk-minter/components/change/ChangeDonationField.tsx (hunk 6 above)
Comment (austinmilt) [reply src/modules/bulk-minter/components/change/ChangeDonationField.tsx:259]: Looks like it's still there. It's not a duplicate line, but a poorly named variable. Suggested change is just to use a more digestible pattern

**Pull request merged**